Fix name loader problem

pp-mo opened the pull request:
Addresses #3288 

~Draft pending :~ 
  - ~[x] tests for enhanced API~
  - ~[x] whatsnew~

pp-mo commented:
@volcan01010 would you be available to review ?

lbdreyer reviewed (commented):
This looks good to me, I had one small comment though.

It is a bit of a shame there is no extra testing for the name loader so it would be nice to get hold of some test data. Is there a way we could generate some?

pp-mo commented on lib/iris/fileformats/name_loaders.py:
```diff
@@ -571,7 +571,9 @@ def _generate_cubes(
 
         if cell_methods is not None:
-            cube.add_cell_method(cell_methods[i])
+            cell_method = cell_methods[i]
```
I don't think this is correct.  The list `cell_methods` contains a result for each "data array" in the file, which is indexed by `i`.  That there is only one in each case is arranged [elsewhere](https://github.com/SciTools/iris/blob/ae2b59ba3dc68fe54893f4e8f4246e5c19c1a5bf/lib/iris/fileformats/name_loaders.py#L579), and I assume is guaranteed by the actual file format (about which I frankly don't know much).

lbdreyer replied:
Ahh yes I see I was reading that `i` as a `1`. That makes more sense now!

pp-mo commented:
> This looks good to me, I had one small comment though.
> 
> It is a bit of a shame there is no extra testing for the name loader so it would be nice to get hold of some test data. Is there a way we could generate some?

As I explained [above](https://github.com/SciTools/iris/pull/4933#issuecomment-1239675286), we could ask @volcan01010 if he is prepared to release his sample file ??
None of the existing test files seem suitable. 
I did test it on a modified version of one of the existing test files, but the problem is I can only guess as to what is correct within the format.

DPeterK commented on lib/iris/cube.py:
```diff
@@ -2293,10 +2293,23 @@ def cell_methods(self):
+                        f"Cube.cell_methods assigned value includes {cell_method}, "
+                        "which is not an iris.coords.CellMethod."
+                    )
+                    raise ValueError(msg)
```
Assumedly we could attempt to handle the incorrectly-typed object here rather than immediately erroring? That way we avoid crashing out of the load at the first sign of failure.

For example, we could instead:
* ignore the not-a-`CellMethod` and not include it in the final list of cell methods. This might mean we don't fully represent all of the averaging metadata of the incoming dataset, of course, which is undesirable - but if the object is already malformed maybe we're better off skipping it anyway?
* try passing the not-a-`CellMethod` back through the `CellMethod` constructor and re-inserting it into the list of cell methods. That way if there is a problem with the averaging metadata we can hopefully fail within the `CellMethod` constructor with a more accurate description of _why_ the object is causing the problem.

pp-mo replied:
> objects that aren't cell methods that show up in the list of cell methods

I don't quite get this.  The change here only operates when you assign something to 'cube.cell_methods', so I can't see why we would want to attempt anything _except_ to say that we only accept CellMethod instances.

My thinking is, I don't think this would ever be triggered within other Iris operations, it would only happen in a user-code assignment.

trexfeathers replied:
I'm happy with this. The main thing is that we're confident the various loaders have already simplified things to either `None` or a valid `CellMethod` - nothing else is expected to get this far unless a user is explicitly using the `setter`.

And in all cases I would be uncomfortable with attempting a conversion to a valid `CellMethod` object, as this is fairly opaque behaviour that separates the user from the true situation (often referred to as "magic").

DPeterK reviewed (commented):
Just one comment from me on handling objects that aren't cell methods that show up in the list of cell methods. To be clear, I don't think this is something that needs to be fixed, but we could consider an alternative approach to erroring fast in the middle of a dataset load.

pp-mo commented:
> None of the existing test files seem suitable. I did test it on a modified version of one of the existing test files, but the problem is I can only guess as to what is correct within the format.

Well, on a closer look I think we actually did have a suitable testfile all along, which was previously unused.
So I used that -- can someone please confirm that it covers the case ?

Meanwhile, apologies for my moving this + some other tests into slightly a more logical hierarchy, it just seemed time to get that done.

pp-mo commented:
OK, I confused myself there.
There is no such pre-existing file in iris-test-data, only in my version !

I will fix this,
 and perhaps now revert all the file moves as it makes the changed-files quite unreasonably messy ...

pp-mo commented:
This should pass, when https://github.com/SciTools/iris-test-data/pull/82 is merged
(unless some other update is merged there first)

bsherratt commented on lib/iris/tests/test_name.py:
```diff
@@ -48,6 +51,32 @@ def test_NAMEII_trajectory(self):
             cubes, ("name", "NAMEIII_trajectory.cml"), checksum=False
         )
 
+    def test_NAMEIII__no_time_averaging(self):
+        cubes = iris.load(
+            tests.get_data_path(("NAME", "NAMEIII_no_time_averaging.txt"))
+        )
```
Actually, the example file is NAME II format, not NAME III.

The first line of the file that says "NAME III" is in fact describing the version of NAME that *produced* the file, not the version of the output. It can do both versions for compatibility.

I think the main way to tell the difference in output formats - certainly the one iris already uses - is whether it has a "Title" (NAME II) or a "Run name" (NAME III).

bsherratt reviewed (commented):
One minor comment about the naming here, but more importantly:
- Having mentioned both NAME II and NAME III formats, they are both widely used in practice (due to a habit of copying NAME config without necessarily reviewing options like the output format), so I would suggest testing both. I can find/create some example output if needed.
- Can this PR please also be added as patch releases to both 3.2 and 3.3? The refactoring of the cube string representation code in 3.2 combines with this bug to mean we often cannot print a cube. While I believe we could do almost everything else with it (except save as originally raised in the linked issue), the lack of printing does not give anyone the confidence to try, so we are essentially stuck using older versions of iris for NAME.

trexfeathers commented:
>Can this PR please also be added as patch releases to both 3.2 and 3.3? The refactoring of the cube string representation code in 3.2 combines with this bug to mean we often cannot print a cube

Offline discussion of the specifics: a `v3.3.1` bugfix would serve, so that's what we will aim for.

volcan01010 commented:
I've had a look through the PR and it looks good to me.  I haven't worked on that Iris project for a while, so unfortunately I'm not up to speed to do a more detailed review.  Please feel free to incorporate the test data file into the project.

pp-mo commented:
Many thanks @bsherratt 
Working slightly in the dark here, as don't actually know much about NAME, so help is welcome !

> Can this PR please also be added as patch releases to both 3.2 and 3.3? The refactoring of the cube string representation code in 3.2 combines with this bug to mean we often cannot print a cube. While I believe we could do almost everything else with it (except save as originally raised in the linked issue), the lack of printing does not give anyone the confidence to try, so we are essentially stuck using older versions of iris for NAME.

From internal discussion (peloton meeting), it seems this really is needed, so yes, we will commit to a 3.3.1 release.
(See #4976 for the beginnings)

----
> Actually, the example file is NAME II format, not NAME III.

Ok I will rename the new file + fix it here and in https://github.com/SciTools/iris-test-data/pull/82

----
> * Having mentioned both NAME II and NAME III formats, they are both widely used in practice (due to a habit of copying NAME config without necessarily reviewing options like the output format), so I would suggest testing both. I can find/create some example output if needed.

This sounds a useful improvement, but we should handle it as a separate issue ?
https://github.com/SciTools/iris/issues/4979
I will contact you offline about getting some suitable testfiles ...

pp-mo commented:
Yay, GTG  !
with thanks, Merges invited ? ...

trexfeathers commented on docs/src/whatsnew/latest.rst:
```diff
@@ -57,6 +57,13 @@ This document explains the changes made to Iris for this release
    variables and cell measures that map to a cube dimension of length 1 are now
    included in the respective vector sections. (:pull:`4945`)
 
+#. `@pp-mo`_ fixed a bug in NAME loaders where data with no associated statistic would
+   load as a cube with invalid cell-methods, which cannot be printed or saved to netcdf.
+   (:issue:`3288`, :pull:`4933`)
+
+#. `@pp-mo`_ ensured that :data:`iris.cube.Cube.cell_methods` must always be an iterable
+   of :class:`iris.coords.CellMethod` objects (:pull:`4933`).
+
```
If you are re-targeting `v3.3.x`, then this entry should be in `3.3.rst`.

trexfeathers reviewed (changes requested):
Thanks @pp-mo! Some tiny testing changes for you, also:

## Don't forget to re-target to `v3.3.x`

pp-mo commented:
> self.assertEqual and self.assertRaisesRegex can be converted to pytest style

Ok I have done that now + I think it looks pretty neat. https://github.com/SciTools/iris/pull/4933/commits/f6853deb1fa2695975fa049f24b3d40f92ff511e

However, It may be appropriate for @lbdreyer  to also check out + OK my use of the fixture here ...

So... I have defined a fixture as a class-instance-method, and labelled it "autouse=True", and it assigns data to "self" properties.
Which is really close to the unittest "setUp" approach -- so I just hope that this is appropriate !!

It seem to work, and I **_think_** it makes sense :  on debugging, I find that the 'self' instance of the "Test__cell_methods" class which the fixture receives is unique for each test method invocation.  So I assume that this 'self' therefore "*is* the test instance", and can be used to hold per-instance data as in the older style.
However, I'd just like a bit more confirmation that this makes sense to someone else, as I don't see this suggestion in the Pytest docs and basically got here more by trial-and-error than from the docs !!

pp-mo commented:
Meanwhile ...
>  Don't forget to re-target to `v3.3.x`

Good spot.
However, I think it's best to let #4976 go first, as that establishes the whatsnew section for 3.3.1.
Rather than changing this + making a conflict, let's wait for that, after which I will fix/rebase this one to build on that accordingly.

pp-mo commented:
Thanks @trexfeathers 
I have now **re-targetted and fixed, and I think will now be OK.**
That includes fixing [the outstanding problem with the post-3.3 whatsnew](https://github.com/SciTools/iris/pull/4976#issuecomment-1257683528) : added an indent to put it inside the v.3.3.1 dropdown.

But I've had to squash commits to sort out the changes :disappointed: 

There was also some suggestion we might want to add the netcdf version pin from https://github.com/SciTools/iris/pull/4968 onto 3.3.x, as otherwise we will probably get occasional docs build failures.
But I'm hoping we can address that separately.

trexfeathers reviewed (approved):
So GitHub is telling me you've changed pretty much everything since my review on Friday, but I can only see the changes I was expecting so I'm assuming that's just the force-pushing messing things up.

lbdreyer commented:
So the docs build is failing again (see https://readthedocs.org/projects/scitools-iris/builds/ for the latests builds) I restarted the job a couple of times but it's getting the same HDF error. Interestingly when @pp-mo [re ran the build ](https://github.com/SciTools/iris/pull/4976#issuecomment-1257683528) it worked the second time

trexfeathers commented:
> So the docs build is failing again (see https://readthedocs.org/projects/scitools-iris/builds/ for the latests builds) I restarted the job a couple of times but it's getting the same HDF error. Interestingly when @pp-mo [re ran the build ](https://github.com/SciTools/iris/pull/4976#issuecomment-1257683528) it worked the second time

>There was also some suggestion we might want to add the netcdf version pin from https://github.com/SciTools/iris/pull/4968 onto 3.3.x, as otherwise we will probably get occasional docs build failures.
>But I'm hoping we can address that separately.

pp-mo commented:
Belated thanks @trexfeathers 
This was one of those occasions when I couldn't get a straight rebase to work for me.
It is unavoidably tough when a rebase/squash interrupts the comment/response cycle, but I needed
Thanks for wading through !
⭐ 🥇

trexfeathers commented:
>This was one of those occasions when I couldn't get a straight rebase to work for me.

@pp-mo next time I'm reviewing feel free to merge in the updates instead of rebasing. It doesn't upset my reviewing workflow and would be easier on you for resolving conflicts.
